evm/chains/sharding/mainchain_handler/vmc_utils.py

```diff
@@ -0,0 +1,212 @@
+import os
+
+import rlp
+
+from viper import compiler
+
+from eth_abi import (
+    decode_abi,
+    encode_abi,
+)
+
+from eth_keys import KeyAPI
+
+import eth_utils
+
+from evm.utils.address import generate_contract_address
+
+from evm.vm.forks.spurious_dragon.transactions import (
+    SpuriousDragonTransaction,
+    SpuriousDragonUnsignedTransaction,
+)
+
+from evm.chains.sharding.mainchain_handler.config import GASPRICE
+
+sha3 = eth_utils.crypto.keccak
+
+WITHDRAW_HASH = sha3("withdraw")
+
+_valmgr_abi = None
+_valmgr_ct = None
+_valmgr_code = None
+_valmgr_bytecode = None
+_valmgr_addr = None
+_valmgr_sender_addr = None
+_valmgr_tx = None
+
+# old one, address=0xCb969cAAad21A78a24083164ffa81604317Ab603
+viper_rlp_decoder_tx = rlp.decode(eth_utils.decode_hex("0xf90237808506fc23ac00830330888080b902246102128061000e60003961022056600060007f010000000000000000000000000000000000000000000000000000000000000060003504600060c082121515585760f882121561004d5760bf820336141558576001905061006e565b600181013560f783036020035260005160f6830301361415585760f6820390505b5b368112156101c2577f010000000000000000000000000000000000000000000000000000000000000081350483602086026040015260018501945060808112156100d55760018461044001526001828561046001376001820191506021840193506101bc565b60b881121561014357608081038461044001526080810360018301856104600137608181141561012e5760807f010000000000000000000000000000000000000000000000000000000000000060018401350412151558575b607f81038201915060608103840193506101bb565b60c08112156101b857600182013560b782036020035260005160388112157f010000000000000000000000000000000000000000000000000000000000000060018501350402155857808561044001528060b6838501038661046001378060b6830301830192506020810185019450506101ba565bfe5b5b5b5061006f565b601f841315155857602060208502016020810391505b6000821215156101fc578082604001510182826104400301526020820391506101d8565b808401610420528381018161044003f350505050505b6000f31b2d4f"), SpuriousDragonTransaction)
```

Comment:
Is there a reason for storing this data as an encoded transaction rather than just it's fields.  It's going to be problematic to hard-code references to the `SpuriousDragonTransaction` since each new set of fork rules will tend to introduce a new transaction rlp class.

Would it work to instead store the individual field values?  Alternatively, we can rlp decode this into a list and just pull the parameters from that list.

Reply:
Actually it is because the `viper_rlp_decoder_tx` is hardcoded in [viper](https://github.com/ethereum/viper/blob/master/viper/utils.py#L106-L111).
Is it help to accept the transaction type as the input to decode each hardcoded transaction and to create the vmc tx?

Reply:
I'll dig into this again on Monday.  I suspect that we can figure out a way to structure this so that it doesn't create a hard dependency on a specific RLP object.

Reply:
No problem. Thank you for pointing it out. For the hardcoded RLP object, I also opened an [issue](https://github.com/ethereum/viper/issues/522) in viper before. Maybe we can also figure out what is the better way?

Reply:
And for the hard-code to `SpuriousDragonTransaction`, is it also happening in https://github.com/ethereum/py-evm/pull/206/files#diff-c1cdc70a34f7db7292b7adabab409d99R122 ?
If yes, is it helping that we have inject the dependency from outside, i.e. make the Transaction class as a parameter for `create_valmgr_tx`?
like
```python
+def create_valmgr_tx(gasprice=GASPRICE, TransactionClass):
    ...
    tx = TransactionClass(0, gasprice, 3000000, b'', 0, bytecode, v, r, s)
    ...
```